src/shogun/classifier/svm/NewtonSVM.cpp

```diff
@@ -30,45 +30,31 @@ NewtonSVM::NewtonSVM() : IterativeMachine<LinearMachine>()
 	t = 0;
 }
 
-NewtonSVM::NewtonSVM(
-    float64_t c, std::shared_ptr<DotFeatures> traindat, std::shared_ptr<Labels> trainlab, int32_t itr)
+NewtonSVM::NewtonSVM(float64_t c, int32_t itr)
     : IterativeMachine<LinearMachine>()
 {
 	lambda=1/c;
 	num_iter=itr;
 	prec=1e-6;
 	C=c;
 	t = 0;
-	set_features(std::move(traindat));
-	set_labels(std::move(trainlab));
 }
 
 
 NewtonSVM::~NewtonSVM()
 {
 }
 
-void NewtonSVM::init_model(std::shared_ptr<Features> data)
+void NewtonSVM::init_model(const std::shared_ptr<Features>& data)
 {
-	if (data)
-	{
-		if (!data->has_property(FP_DOT))
-			error("Specified features are not of type CDotFeatures");
-		set_features(std::static_pointer_cast<DotFeatures>(data));
-	}
-
-	ASSERT(features)
-
-	SGVector<float64_t> train_labels = binary_labels(m_labels)->get_labels();
+	const auto features = data->as<DotFeatures>();
```

Comment:
@gf712 you think it might be worth having a check for these in the base class as well?

Reply:
yes agreed. You mean check this in `LinearMachine` right? And the here it's a static cast?

Reply:
while I agree it could be a static cast in a perfect world, in this world it seems unsafe (in case we forget something) ... but I would just make sure to catch as much as possible in base classes

Reply:
Do all LinearMachine algorithms use `DotFeatures`? If so we could just write train_machine as final in LinearMachine and then dispatch the call to another virtual function for LinearMachine implementations, e.g. `LinearMachine::iteration(const std::shared_ptr<DotFeatures>& data, const std::shared_ptr<Labels>& labs)`

Reply:
something like this https://godbolt.org/z/q8QsAs

Reply:
I see, make sense, I will do this.

Reply:
makes sense!